components/fxa-client/src/oauth.rs

```diff
@@ -163,20 +163,67 @@ impl FirefoxAccount {
     /// * `state` - OAuth state.
     /// * `access_type` - Type of OAuth access, can be "offline" and "online.
     pub fn authorize_code_using_session_token(
-        &self,
+        // Is mut purely for the sake of this draft
+        // Will be removed
+        &mut self,
         client_id: &str,
         scope: &str,
         state: &str,
         access_type: &str,
     ) -> Result<String> {
         let session_token = self.get_session_token()?;
+        let mut code_challenge = None;
+        let mut keys_jwe = None;
+        let mut code_challenge_method = None;
+        // Is this check even neccesary? Can we just compute the keys_jwe and
+        // Send it anyway, filtering the self.state.scoped_keys on the reqeusted scopes
+        if scope
+            .split_whitespace()
+            .any(|split_scope| split_scope == scopes::OLD_SYNC)
+        {
+            let code_verifier = util::random_base64_url_string(43)?;
+            let code_challenge_digest = digest::digest(&digest::SHA256, &code_verifier.as_bytes())?;
```

Comment:
Generating these here does not feel right to me (but I could be misunderstanding or misremembering what this is trying to achieve).

In my head, we're expanding this method so that we can use it for authorizing *some other client* to access the user's account, e.g. as part of the pairing flow. In that scenario, we're conceptually acting as part of the FxA server and fulfilling an OAuth request that we've received from some other client.

If that's the case, then generating `code_challenge` here ourselves doesn't achieve anything. When we get the OAuth code from the server and send it back to the requesting client, that client won't be able to use it because it doesn't know the corresponding `code_verifier`.

I expect that we will need to accept `code_challenge`, `code_challenge_method` and `keys_jwk` as input parameters to this function, rather than generating them ourselves internally. (And you may find that doing this, actually makes the integration code easier, because you don't have to mess with `self.flow_store`).

Reply:
Alrighty updated it to do that. Makes a lot more sense. I also propagated the parameters all the way up the FFI so the caller can set the parameters. This would then be a breaking change 😅 (Although it seems like this API is only used in Android Components? I could make a PR for them if this ends up merging in this state)